app/src/main/java/org/mozilla/fenix/utils/BrowsersCache.kt

```diff
@@ -0,0 +1,54 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.utils
+
+import android.content.Context
+import androidx.annotation.VisibleForTesting
+import mozilla.components.support.utils.Browsers
+
+/**
+ * Caches the list of browsers installed on a user's device.
+ *
+ * BrowsersCache caches the list of installed browsers is gathered lazily when it is first accessed
+ * after initial creation or invalidation. For that reason, a context is required every time
+ * the cache is accessed.
+ *
+ * In addition, a context is required every time the cache is accessed because every access
+ * is guaranteed to return cached results only if the context parameter of the access matches the
+ * context parameter given when the cache was filled. This ensures that if the cache is filled with
+ * a list of browsers installed for a particular context, that list will be invalidated and replaced
+ * if the cache is accessed for a different context.
+ *
+ * Users are responsible for invalidating the cache at the appropriate time. It is left up to the
+ * user to determine appropriate policies for maintaining the validity of the cache. As mentioned
+ * above, the cache will be lazily refilled after invalidation. In other words, invalidation is O(1).
+ */
+object BrowsersCache {
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal var cachedBrowsers: Browsers? = null
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal var cachedContext: Context? = null
+
+    @Synchronized
+    fun all(context: Context): Browsers {
+        run {
```

Comment:
nit: why is this in a run block? It seems unnecessary and adds a potential source of confusion since `run` returns the last expression so `return cachedBrowsers` is actually escaping that scoping to the parent function, which may be unclear.

Reply:
Not unclear. It's exactly what I want to happen. I am basically using it as a scoping operator.

Reply:
I don't think it's idiomatic in Kotlin – which can make it non-obvious to devs that it's a scoping operator – but wfm.

I wonder if a set of braces (also non-idiomatic but maybe more obvious) can actually be used as a scoping operator too... Or is that an anonymous function? lol

Reply:
The IDE yells at my for braces: it's an anonymous function.